File: Problems/python/boj_단어뒤집기2.py
#https://www.acmicpc.net/problem/17413
#implementation

s = list(input()) #str은 부분 수정 불가능 하므로 list로 바꿔서 사용
s.append(' ') #마지막 idx 사용하고 싶으면 적절한 요소 추가

tag = False
tmp = []

for idx, i in enumerate(s):
    if i=='<' or idx==len(s)-1:
        tag = True

        if tmp: #tmp에 내용이 있으면
            tmp = ''.join(tmp).split()
            ptmp = ''
            for j in tmp:
                ptmp += j[::-1] + ' ' #뒤집어서
            s[idx-len(ptmp)+1:idx] = ptmp[:-1] #s에 저장
            tmp = []

    elif i == '>':
        tag = False
        continue #그대로 넘어가면 tmp에 '>'도 저장됨.

    if not tag: #뒤집어야할 것만 tmp에 저장
        tmp.append(i)

print(''.join(s)) #list인 s를 str로 출력

File: Problems/cpp/boj_단어뒤집기2.cpp
//https://www.acmicpc.net/problem/17413
//Implementation

#include <iostream>
#include <string>
#include <sstream>

using namespace std;

int main(){
    ios::sync_with_stdio(0);
    cin.tie(0);
    
    string str;
    getline(cin, str);
    str += ' ';

    bool reverse = false;
    string tmp = "";
    for (int i=0; i<str.size(); i++){
        if (str[i]=='<' || i==str.size()-1){
            if (tmp.size()){
                string rtmp = "";
                stringstream ss(tmp);
                string token;
                while (getline(ss, token, ' ')){
                    rtmp += string(token.rbegin(), token.rend());
                    rtmp += ' ';
                }
                rtmp.pop_back();
                cout << rtmp;
            }

            tmp = "";
            reverse = true;
        }
        else if (str[i]=='>'){
            reverse = false;
            cout << str[i];
            continue;
        }

        if (reverse){
            cout << str[i];
        }
        else{
            tmp += str[i];
        }
    }
}

//문자열을 하나씩 출력하면서 진행할지
//원본 문자열의 요소들을 바꿔서 마지막에 한번에 출력할지.